### ISSUE TITLE
Docgen: Crash for untyped @param descriptions on default re-export

### ISSUE THREAD

**Issue by dmsnell** [WordPress/gutenberg#37766]
### Description

In context-provider modules we typically re-export the `Provider` as the default export, wrapping a custom context.

```js
export const Context = createContext( false );

const { Consumer, Provider } = Context;

/**
 * @param {boolean} props.value Whether to turn the thing on.
 * @returns {WPComponent} Provider component
 */
export default Provider;
```

For TypeScript modules we don't need or want the type annotations as those are enforced and maintained by the compiler.

```ts
/**
 * @param props.value Whether to turn the thing on.
 * @returns Provider component
 */
export default Provider;
```

When generating the docs though it fails with a couple of different types of errors.

```
TypeError: Cannot read properties of undefined (reading 'name')
```

```
TypeError: Cannot read properties of undefined (reading '0')
```

These come from trying to [get the first parameter](https://github.com/WordPress/gutenberg/blob/trunk/packages/docgen/lib/get-type-annotation.js#L460) and [printing the name of the function in the error log](https://github.com/WordPress/gutenberg/blob/trunk/packages/docgen/lib/get-type-annotation.js#L467).

### Step-by-step reproduction instructions

See Description

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by sarayourfriend** [comment]
I think this is a case of incorrectly using JSDoc TypeScript annotations and also a misunderstanding about what it possible with the current docgen TypeScript implementation.

The `@param` and `@returns` annotations are only valid for a function declaration. Re-exporting a variable should use the `@type` annotation (I think) like so:

```js

/**
 * @type {(props: { value: boolean }) => import('react').Component}
 */
export default Provider;
```

Even though it's TypeScript, we cannot use the type information from TypeScript in this case. It is an inherit limitation to the way docgen works. We don't actually run the TypeScript compiler to retrieve type information for exported members, we statically analyse the TypeScript annotations in the TypeScript source code itself.

One potential workaround is to explicitly annotate the type of the re-exported member:

```ts
export const Provider: (props: { value: boolean }) => import('react').Component;
```

That's not ideal, but it's the only way we can do it at the moment without significantly re-writing how docgen handles TypeScript files (which would require generally also re-writing the upper level docgen code as well, probably to create some sort of plugin system, which it was not originally designed for—this would not be trivial).

**Comment by sarayourfriend** [comment]
Oh I'll add the above with the caveat that even if we can't support this usage, we should emit a sensible warning instead of crashing. At the very least this issue should address that.

**Comment by dmsnell** [comment]
> a misunderstanding about what it possible with the current docgen TypeScript implementation

I'm worried that the limitations of the current implementation are going to bite us over and over again, specifically the way it reports on types. This probably encroaches on bigger discussions about the intersection of TypeScript types, descriptions of parameters/properties, and our published documentation.

The docgen is going to have issues with an increasing number of normative code patterns as we transition more modules into TypeScript. Here are a couple more examples from today:

```md
`ReturnType< >`: Higher-order component.
```

That type threaded through the return type of the function it called. `docgen` discarded the most important part of the type but it didn't complain.

```md
( mapComponent: HigherOrderComponent< HOCProps >, modifierName: string ): typeof mapComponent {
```

Here's another case where it didn't complain. Initially `typeof mapComponent` wasn't there and the return type was inferred. Putting aside the discussion of whether public functions should be allowed to omit the return type annotation, the code was fine and TypeScript had a type (albeit a complicated one that probably isn't that helpful in the docs anyway `<InnerProps extends HOCProps>(Inner: ComponentType<InnerProps>) => {} extends HOCProps ? ComponentType<...> : ComponentType<...>`) but the `docgen` crashed again saying it couldn't find the type.

---

Apart from calling TypeScript to give us the types (an idea I think may also be build on wrong assumptions) do you think there's something we could do to prevent the auto-generation of type information in the docs from blocking changes? Could we have it link to the code if it fails to get a type? Omit the type entirely?

The stronger our TypeScript code is the fewer type annotations I think it will have; that's always one of my top goals at least since it provides all the benefit with a minimal amount of inconvenience. There's a big mismatch between the way we build our docs and that reality though.

**Comment by sarayourfriend** [comment]
> Omit the type entirely?

Yes I think this is the right solution in these cases. Produce less documentation, not more, _incorrect_ documentation (or fail to finish producing it altogether).

> The stronger our TypeScript code is the fewer type annotations I think it will have; that's always one of my top goals at least since it provides all the benefit with a minimal amount of inconvenience. There's a big mismatch between the way we build our docs and that reality though.

I think that's why moving towards something like TypeDoc eventually is a good thing. We'll get the benefit of those strong types not just when writing and verifying code, but also when learning about new areas from the written documentation.

**Comment by dmsnell** [comment]
> I think that's why moving towards something like TypeDoc eventually is a good thing.

Generating documentation from comments inside the types and interfaces would be a good thing. That would restore some of the mismatch between documented JSDoc types and TypeScript types. It does appear to support JSDoc-style tags.

Do we know how much people are using these? How much effort we're saving by auto-generating them? As the types get more complicated I wonder about how useful they will actually be to developers vs. having a description that explains their usefulness.

---

If I can make some time within the next week or so I'll see if I can have it abort on "no type inferred" to clear up this particular issue. Thanks for discussing it!

**Comment by sarayourfriend** [comment]
I believe it is used to generate the Gutenberg handbook, so presumably quite a good deal of people are using it.

**Comment by dmsnell** [comment]
> I believe it is used to generate the Gutenberg handbook, so presumably quite a good deal of people are using it.

My mistake; I wasn't talking about the docs in general, but specifically the type annotations. For instance, they are littered with `Object` and `Function` which don't say much and leave people still in the position to have to read the source to find out what the expectations are. I was wondering more if you've heard of people relying on those type listings in the docs. Would be hard to find via metrics on account of the fact that you can't view the docs page without also seeing the types.

**Comment by sarayourfriend** [comment]
Right, I have no idea. I can share my anecdotal experience with the handbook when I first started to learn Gutenberg development roughly two years ago... it was borderline useless for me, and reading the source code was the only way I found that helped me understand how to use stuff or what anything was doing. You're definitely right about `Object` and `Function` types being more or less useless at conveying any interesting or useful information to someone using the docs as a reference tool.

Update: I'll clarify that I was reading the handbook for things like wp/data and wp/components. The handbook's descriptions of other packages are usually better in my experience, like the various script and bootstraping packages for creating blocks.

**Comment by dmsnell** [comment]
If nothing else, your experience then seems to match mine. I'm guessing nobody would miss it that much if we stripped all type annotations from the docs…

**Comment by sarayourfriend** [comment]
We could publish something on https://make.wordpress.org/core/ asking how folks feel about removing the types from the documentation.

cc @gziolo who I think would probably have a gut sense of how that would go over with folks.

**Comment by gziolo** [comment]
I personally have no idea how much useful types are for folks browsing the block editor handbook, but you see only text so anything more complex than primitives is hard to follow anyway. If you want to find out we can discuss it during the bi-weekly Core JS office hours and check whether we go as far as publishing at https://make.wordpress.org/core/.
